File: internal/apps/diagnostics/normalize.go
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package diagnostics

import (
	"crypto/sha256"
	"encoding/hex"
	"regexp"
	"strings"
)

var (
	exceptionClassPattern   = regexp.MustCompile(`([A-Za-z0-9_$.]+(?:Exception|Error))`)
	javaThreadPrefixPattern = regexp.MustCompile(`^Exception in thread "[^"]+"\s+`)
	seatunnelLogHeaderPattern = regexp.MustCompile(`^(?:\[[^\]]*\]\s+)?\d{4}-\d{2}-\d{2}[ T]\d{2}:\d{2}:\d{2}(?:[\.,]\d{3})?\s+(?:TRACE|DEBUG|INFO|WARN|ERROR|FATAL)\s+(?:\[[^\]]*\]\s+){0,2}-\s*`)
	timestampPattern        = regexp.MustCompile(`\d{4}-\d{2}-\d{2}[ T]\d{2}:\d{2}:\d{2}(?:[\.,]\d{3})?`)
	uuidPattern             = regexp.MustCompile(`[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}`)
	ipPattern               = regexp.MustCompile(`\b\d{1,3}(?:\.\d{1,3}){3}\b`)
	hexPattern              = regexp.MustCompile(`0x[0-9a-fA-F]+`)
	durationPattern         = regexp.MustCompile(`\b\d+(?:\.\d+)?\s*(?:ns|us|µs|ms|s|m|h|sec|secs|second|seconds|minute|minutes|hour|hours)\b`)
	numberPattern           = regexp.MustCompile(`\b\d+\b`)
	statusSuffixPattern     = regexp.MustCompile(`\b([A-Z][A-Z0-9_]{2,}:\s+.+)$`)
	separatorOnlyPattern    = regexp.MustCompile(`^[=\-_*#~\s]+$`)
	whitespacePattern       = regexp.MustCompile(`\s+`)
)

var diagnosticNoiseExactLines = map[string]struct{}{
	"fatal error": {},
	"please submit bug report in https://github.com/apache/seatunnel/issues": {},
	"reason:seatunnel job executed failed":                                   {},
}

// BuildErrorFingerprint normalizes a Seatunnel error sample into a stable fingerprint.
// BuildErrorFingerprint 将 Seatunnel 错误样本归一化为稳定指纹。
func BuildErrorFingerprint(message, evidence string) (fingerprint, normalized, exceptionClass, title string) {
	raw := strings.TrimSpace(evidence)
	if raw == "" {
		raw = strings.TrimSpace(message)
	}
	rootLine := extractRootCauseLine(message, raw)
	exceptionClass = ExtractExceptionClass(rootLine)
	title = buildErrorTitle(rootLine, message, raw, exceptionClass)
	normalized = normalizeFingerprintText(strings.Join(composeFingerprintLines(title, rootLine, raw, exceptionClass), "\n"))
	if normalized == "" {
		normalized = normalizeFingerprintText(title)
	}
	if normalized == "" {
		return "", "", "", ""
	}
	hash := sha256.Sum256([]byte(normalized))
	return hex.EncodeToString(hash[:]), normalized, exceptionClass, title
}

// ExtractExceptionClass extracts the first exception/error class from text.
// ExtractExceptionClass 从文本中提取第一个异常/错误类名。
func ExtractExceptionClass(value string) string {
	matches := exceptionClassPattern.FindStringSubmatch(value)
	if len(matches) != 2 {
		return ""
	}
	return strings.TrimSpace(matches[1])
}

func buildErrorTitle(rootLine, message, evidence, exceptionClass string) string {
	candidates := []string{
		strings.TrimSpace(rootLine),
		strings.TrimSpace(message),
		firstMeaningfulLine(evidence),
	}
	for _, item := range candidates {
		item = canonicalizeFingerprintLine(item)
		if item == "" || isDiagnosticNoiseLine(item) {
			continue
		}
		if exceptionClass != "" && !strings.Contains(item, exceptionClass) {
			return exceptionClass + ": " + item
		}
		return item
	}
	return exceptionClass
}

func composeFingerprintLines(title, rootLine, evidence, exceptionClass string) []string {
	result := make([]string, 0, 6)
	seen := make(map[string]struct{}, 8)
	appendUnique := func(value string) {
		value = canonicalizeFingerprintLine(value)
		if isDiagnosticNoiseLine(value) {
			return
		}
		if value == "" {
			return
		}
		if _, exists := seen[value]; exists {
			return
		}
		seen[value] = struct{}{}
		result = append(result, value)
	}

	appendUnique(exceptionClass)
	appendUnique(rootLine)
	if len(result) > 0 {
		return result
	}

	appendUnique(title)

	for _, line := range strings.Split(evidence, "\n") {
		trimmed := strings.TrimSpace(line)
		if trimmed == "" || strings.HasPrefix(trimmed, "at ") {
			continue
		}
		appendUnique(trimmed)
		if len(result) >= 6 {
			break
		}
	}
	return result
}

func normalizeFingerprintText(value string) string {
	normalized := strings.ToLower(strings.TrimSpace(value))
	normalized = timestampPattern.ReplaceAllString(normalized, "<ts>")
	normalized = uuidPattern.ReplaceAllString(normalized, "<uuid>")
	normalized = ipPattern.ReplaceAllString(normalized, "<ip>")
	normalized = hexPattern.ReplaceAllString(normalized, "<hex>")
	normalized = durationPattern.ReplaceAllString(normalized, "<duration>")
	normalized = numberPattern.ReplaceAllString(normalized, "<num>")
	normalized = whitespacePattern.ReplaceAllString(normalized, " ")
	return strings.TrimSpace(normalized)
}

func extractRootCauseLine(message, evidence string) string {
	if causedBy := lastCausedByLine(evidence); causedBy != "" {
		return canonicalizeFingerprintLine(causedBy)
	}
	for _, candidate := range []string{
		strings.TrimSpace(message),
		firstMeaningfulLine(evidence),
	} {
		if line := canonicalizeFingerprintLine(candidate); line != "" && !isDiagnosticNoiseLine(line) {
			return line
		}
	}
	return ""
}

func lastCausedByLine(value string) string {
	result := ""
	for _, line := range strings.Split(value, "\n") {
		trimmed := strings.TrimSpace(line)
		if strings.HasPrefix(trimmed, "Caused by:") {
			candidate := strings.TrimSpace(strings.TrimPrefix(trimmed, "Caused by:"))
			if !isDiagnosticNoiseLine(candidate) {
				result = candidate
			}
		}
	}
	return result
}

func canonicalizeFingerprintLine(value string) string {
	trimmed := strings.TrimSpace(value)
	if trimmed == "" {
		return ""
	}
	if strings.HasPrefix(trimmed, "Caused by:") {
		trimmed = strings.TrimSpace(strings.TrimPrefix(trimmed, "Caused by:"))
	}
	if strings.HasPrefix(trimmed, "Exception StackTrace:") {
		trimmed = strings.TrimSpace(strings.TrimPrefix(trimmed, "Exception StackTrace:"))
	}
	trimmed = strings.TrimSpace(seatunnelLogHeaderPattern.ReplaceAllString(trimmed, ""))
	trimmed = strings.TrimSpace(javaThreadPrefixPattern.ReplaceAllString(trimmed, ""))
	if matches := statusSuffixPattern.FindStringSubmatch(trimmed); len(matches) == 2 {
		trimmed = strings.TrimSpace(matches[1])
	}
	trimmed = whitespacePattern.ReplaceAllString(trimmed, " ")
	return strings.TrimSpace(trimmed)
}

func firstMeaningfulLine(value string) string {
	for _, line := range strings.Split(value, "\n") {
		trimmed := strings.TrimSpace(line)
		if trimmed == "" || strings.HasPrefix(trimmed, "at ") {
			continue
		}
		candidate := canonicalizeFingerprintLine(trimmed)
		if isDiagnosticNoiseLine(candidate) {
			continue
		}
		return candidate
	}
	return ""
}

func isDiagnosticNoiseLine(value string) bool {
	trimmed := strings.TrimSpace(value)
	if trimmed == "" {
		return true
	}
	if separatorOnlyPattern.MatchString(trimmed) {
		return true
	}
	normalized := strings.ToLower(strings.Trim(trimmed, " \t\r\n,.;:-"))
	if normalized == "" {
		return true
	}
	_, ok := diagnosticNoiseExactLines[normalized]
	return ok
}

File: frontend/lib/services/st-upgrade/normalize.ts
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import type {
  ConfigMergeFile,
  ConfigMergePlan,
  PrecheckResult,
  TaskListData,
  TaskLogsData,
  TaskStepsData,
  UpgradePlanRecord,
  UpgradeStepLog,
  UpgradeTask,
  UpgradeTaskStep,
  UpgradeTaskSummary,
  UpgradeNodeExecution,
} from './types';
import {getCurrentLocale, type Locale} from '@/lib/i18n/config';
import {localizeBackendText} from '@/lib/i18n/localize-text';

export function localizeUpgradeText(
  value?: string | null,
  locale: Locale = getCurrentLocale(),
): string {
  return localizeBackendText(value, locale);
}

function sanitizeConfigMergeFile(file: ConfigMergeFile): ConfigMergeFile {
  const conflicts = Array.isArray(file.conflicts) ? file.conflicts : [];

  return {
    ...file,
    conflicts,
  };
}

export function sanitizeConfigMergePlan(plan?: ConfigMergePlan | null): ConfigMergePlan | null {
  if (!plan) {
    return null;
  }

  const files = Array.isArray(plan.files) ? plan.files.map(sanitizeConfigMergeFile) : [];

  return {
    ...plan,
    files,
  };
}

export function sanitizePrecheckResult(precheck?: PrecheckResult | null): PrecheckResult | null {
  if (!precheck) {
    return null;
  }

  return {
    ...precheck,
    issues: Array.isArray(precheck.issues)
      ? precheck.issues.map((issue) => ({
          ...issue,
          message: localizeUpgradeText(issue.message),
        }))
      : [],
    node_targets: Array.isArray(precheck.node_targets) ? precheck.node_targets : [],
    config_merge_plan: sanitizeConfigMergePlan(precheck.config_merge_plan) || precheck.config_merge_plan,
  };
}

export function sanitizeUpgradePlanRecord(plan?: UpgradePlanRecord | null): UpgradePlanRecord | null {
  if (!plan) {
    return null;
  }

  return {
    ...plan,
    snapshot: {
      ...plan.snapshot,
      node_targets: Array.isArray(plan.snapshot.node_targets) ? plan.snapshot.node_targets : [],
      steps: Array.isArray(plan.snapshot.steps) ? plan.snapshot.steps : [],
      config_merge_plan: sanitizeConfigMergePlan(plan.snapshot.config_merge_plan) || plan.snapshot.config_merge_plan,
    },
  };
}

function sanitizeTaskStep(step: UpgradeTaskStep): UpgradeTaskStep {
  return {
    ...step,
    message: localizeUpgradeText(step.message),
    error: localizeUpgradeText(step.error),
  };
}

function sanitizeNodeExecution(node: UpgradeNodeExecution): UpgradeNodeExecution {
  return {
    ...node,
    message: localizeUpgradeText(node.message),
    error: localizeUpgradeText(node.error),
  };
}

export function sanitizeUpgradeTask(task?: UpgradeTask | null): UpgradeTask | null {
  if (!task) {
    return null;
  }

  return {
    ...task,
    failure_reason: localizeUpgradeText(task.failure_reason),
    rollback_reason: localizeUpgradeText(task.rollback_reason),
    plan: sanitizeUpgradePlanRecord(task.plan) || task.plan,
    steps: Array.isArray(task.steps) ? task.steps.map(sanitizeTaskStep) : [],
    node_executions: Array.isArray(task.node_executions)
      ? task.node_executions.map(sanitizeNodeExecution)
      : [],
  };
}

function sanitizeUpgradeTaskSummary(
  task: UpgradeTaskSummary,
): UpgradeTaskSummary {
  return {
    ...task,
    failure_reason: localizeUpgradeText(task.failure_reason),
    rollback_reason: localizeUpgradeText(task.rollback_reason),
  };
}

export function sanitizeTaskListData(data?: TaskListData | null): TaskListData | null {
  if (!data) {
    return null;
  }

  return {
    ...data,
    items: Array.isArray(data.items)
      ? data.items.map(sanitizeUpgradeTaskSummary)
      : [],
  };
}

export function sanitizeTaskStepsData(data?: TaskStepsData | null): TaskStepsData | null {
  if (!data) {
    return null;
  }

  return {
    ...data,
    steps: Array.isArray(data.steps) ? data.steps.map(sanitizeTaskStep) : [],
    node_executions: Array.isArray(data.node_executions)
      ? data.node_executions.map(sanitizeNodeExecution)
      : [],
  };
}

function sanitizeUpgradeStepLog(log: UpgradeStepLog): UpgradeStepLog {
  return {
    ...log,
    message: localizeUpgradeText(log.message),
  };
}

export function sanitizeTaskLogsData(data?: TaskLogsData | null): TaskLogsData | null {
  if (!data) {
    return null;
  }

  return {
    ...data,
    items: Array.isArray(data.items)
      ? data.items.map(sanitizeUpgradeStepLog)
      : [],
  };
}
